Sources/NIOSSL/SSLCallbacks.swift

```diff
@@ -173,6 +173,117 @@ public struct PSKClientIdentityResponse: Sendable {
     }
 }
 
+/// A structure representing values from client extensions in the SSL/TLS handshake.
+///
+/// This struct contains values obtained from the client hello message extensions during the TLS handshake process and
+/// can be manipulated or introspected by the `NIOSSLContextCallback` to alter the TLS handshake behaviour dynamically
+/// based on these values.
+public struct NIOSSLClientExtensionValues: Hashable {
+
+    /// The hostname value from the Server Name Indication (SNI) extension.
+    ///
+    /// This value, if available, indicates the requested server hostname by the client.
+    /// In a context where a service is handling multiple hostnames (virtual hosts, for example),
+    /// this value could be used to decide which SSLContext to use for the handshake.
+    public var serverHostname: String?
+
+    /// Initializes a new `NIOSSLClientExtensionValues` struct.
+    ///
+    /// - parameter serverHostname: The hostname value from the SNI extension.
+    public init(serverHostname: String?) {
+        self.serverHostname = serverHostname
+    }
+}
+
+/// A structure representing changes to the SSL/TLS configuration that can be applied 
+/// after the client hello message extensions have been processed.
+public struct NIOSSLContextPostExtensionConfigurationChange {
```

Comment:
I think this type looks totally reasonable. Instead of calling it a "change", as it's not a diff, let's call it a "configuration override".

Don't forget to add a `public init()` or users won't be able to construct it. Let's avoid having the init take parameters, just have it default to all-nil.

Reply:
@Lukasa Just pushed this up, but I added the params to constructor (default nil) cause I think its a bit more ergonomic from user perspective to pass in those values if you have them instead of making a new `var` and mutating before returning.

Reply:
While I agree it's more ergonomic, as we add more fields to the structure we have to add new initializers to make them set that way (for API stability reasons we can't add new arguments with default values). That has the effect of making the type explode with values. In this case, as we know we're going to want to add more, we should just start by putting everything in that footing and get people writing the code we expect.

Reply:
Ah thats helpful context, just pushed up empty init. Do you think there's any value in letting user return `nil` in the callback? Obviously can just return an empty `.init()` as well if no changes are needed.

Reply:
Nice just pushed